csrc/cpu/hg_sample.cpp

```diff
@@ -0,0 +1,162 @@
+				col_store,
+				sampled_nodes_store,
+				&budget_store
+			);
+		}
+	}
+
+
+	// Sampling process
+	for (int l = 0; l < num_layers; l++) {	
+		for (const auto &i : _budget_store) {
+			const auto &node_type = i.first;
+			auto &per_type_budget = i.second;
+
+			 vector<int64_t> samples = sample_nodes(per_type_budget, n);
+
+			 // Remove sampled nodes from the budget
+			 for (const auto &sample : samples) {
+			 	per_type_budget.erase(*sample);
+			 }
+
+			 type_to_n_ids.insert(node_type, samples);
+		}
+	}
+
+	// Re-index
+	c10::Dict<string, std::vector<int64_t>> type_to_n_ids;
+
+}
+
+void update_budget(
+	int64_t added_node_idx,
+	const c10::Dict<rel_t, torch::Tensor> &rowptr_store,
+	const c10::Dict<rel_t, torch::Tensor> &col_store,
+	const c10::Dict<node_t, std::set<int64_t>> &sampled_nodes_store,
+	c10::Dict<string, c10::Dict<int64_t, float>> *budget_store,
+) {
+	for (const auto &i : rowptr_store) {
+		const rel_t &relation_type = i.first;
+		int64_t *row_ptr_raw = i.second.data_ptr<int64_t>();
+		int64_t *col_raw = col_store.at(relation_type).data_ptr<int64_t>();
+		
+		// Get the budget map and sampled_nodes for the source node type of the relation
+		const auto &source_node_type = std::get<0>(relation_type);
+		const std::set<int64_t> &sampled_nodes = sampled_nodes_store.at(source_node_type);
+		c10::Dict<int64_t, float> *budget = &budget_store->at(source_node_type);
+		
+		int64_t row_start_idx = row_ptr_raw[added_node_idx];
+		int64_t row_end_idx = row_ptr_raw[added_node_idx + 1];
+		if (row_start_idx != row_end_idx) {
+			// Compute the norm of degree and update the budget for the neighbors of added_node_idx
+			double norm_deg = 1 / (double)(row_end_idx - row_start_idx);
+			for (int64_t j = row_start_idx; j < row_end_idx; j++) {
+				if (sampled_nodes.find(col_raw[j]) == sampled_nodes.end()) {
+					const auto &it = budget->find(col_raw[j]);
+					float val = it != budget->end() ? it.second : 0.0;
+					budget->insert_or_assign(col_raw[j], val + norm_deg);
+				}
+			}
+		}
+	}
+}
+
+// Sample n nodes according to its type budget map. The probability that node i is sampled is calculated by
+// prob[i] = budget[i]^2 / l2_norm(budget)^2.
+vector<int64_t> sample_nodes(const c10::Dict<int64_t, float> &budget, int n) {	
+	// Compute the squared L2 norm	
+	float norm = 0.0;
+	for (const auto &i : budget) {
+		norm += i.second * i.second;	
+	}
+
+	// Generate n sorted random values between 0 and norm
+	std::vector<double> samples(n);
+	std::uniform_real_distribution<double> dist(0.0, norm);
+	std::generate(std::begin(x), std::end(x), [&]{ return dist(gen); });
+	std::sort(samples.begin(), samples.end());
+
+	// Iterate through the budget map to compute the cumulative probability cum_prob[i] for node_i. The j-th
+	// sample is assigned to node_i iff cum_prob[i-1] < samples[j] < cum_prob[i]. The implementation assigns
+	// two iterators on budget and samples respectively, then computes the node samples in linear time by 
```

Comment:
Not sure if I understand that yet. Do there exists some references that describe the sampling strategy? Is it the most efficient way to do it (since it requires a sort)?

Reply:
It basically contains logic to do weighted sample for N elements in O(N log N). Do you know a way to do it faster in O(N) without too much overhead (for example, setting up Alias Sampling might be too heavy-weight).

The current implementation is doing something similar to this to sample N elements:
1. Uniform random sample N samples between [0.0, 1.0]
2. We compute the cumulative probability bucket for each candidate by doing `cand_cum_p[i] = cand_cum_p[i-1] + cand_p[i]`. Basically, if the sample falls between `cand_cum_p[i - 1]` and `cand_cum_p[i]` then we would return i as the sample.
3. Now the problem becomes for each sample in (1), we have to find i such that cand_cum_p[i-1] < sample < cand_cum_p[i]. 
4. Instead of binary search, we can sort the samples from (1) and use the two iterator approach to find the i for each sample. We start iterator i at the beginning of cand_cum_p and j at the beginning of sample. Then we repeat these steps:
    - If sample[j] > cand_cum_p[i], then we can increment i until sample[j] < cand_cum_p[i]. 
    - If sample[j] < cand_cum_p[i], then we can increment j until sample[j] > cand_cum_p[i]. The i is also the sample for each j.

Reply:
Thanks for the clarification. This makes sense.